### PR TITLE
One bug fix and two new features

### PR DESCRIPTION
**Pull request by yu239**
I closed two old pull requests and add them together in this one.
1. rotate_layer and flip_layer. These two operations are performed on each input vector (row of mini-batch matrix)
2. added getMin and getMax for GpuMatrix
3. bug fix for matrix transpose when memalloc is true (the matrix pointer needs to be created)

### CODE REVIEW

File: paddle/math/tests/test_SparseMatrix.cpp [hunk 1]
```diff
@@ -248,11 +248,13 @@ TEST(Matrix, SparseMatrixTranspose) {
...
-            CpuMatrixPtr matD(new CpuMatrix(width, height));
+            MatrixPtr matD(new CpuMatrix(width, height));
             matC->transpose(matD, false);
+
             /*check result*/
             checkSMatrixEqual2Dense(
-                std::dynamic_pointer_cast<CpuSparseMatrix>(matB), matD);
+                std::dynamic_pointer_cast<CpuSparseMatrix>(matB),
+                std::dynamic_pointer_cast<CpuMatrix>(matD));
           }
```
Comment (wangyang59) [inline paddle/math/tests/test_SparseMatrix.cpp]: what is the purpose of this change here?

File: paddle/math/tests/test_matrixCompare.cpp [hunk 2]
```diff
@@ -181,7 +196,7 @@ void testMatrixInverse(int height) {
   cpu->add(*outputCheck);
 
   gpu->copyFrom(*cpu);
-  cpu->inverse(cpuI, false);
+  cpu->inverse(cpuI, true);
   gpu->inverse(gpuI, false);
```
Comment (wangyang59) [inline paddle/math/tests/test_matrixCompare.cpp]: what is the purpose of the change here?

File: paddle/math/Matrix.cpp [hunk 3]
```diff
@@ -1708,13 +1736,35 @@ void CpuMatrix::transpose(MatrixPtr matTrans, bool memAlloc) {
   }
 }
```
Comment (wangyang59) [inline paddle/math/Matrix.cpp]: please add the math description here

File: paddle/gserver/layers/RotateLayer.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+#pragma once
+
+#include "Layer.h"
+#include "paddle/math/Matrix.h"
+
+namespace paddle {
+/**
+ * A layer for rotating an input sample (assume it's a matrix)
+ * The rotation is in clock-wise
```
Comment (yu239) [inline paddle/gserver/layers/RotateLayer.h]: fixed

File: paddle/gserver/layers/RotateLayer.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  explicit RotateLayer(const LayerConfig& config) : Layer(config) {}
+
+  bool init(const LayerMap& layerMap, const ParameterMap& parameterMap);
+
+  void forward(PassType passType);
+  void backward(const UpdateCallback& callback = nullptr);
+
+private:
+  int batchSize_;
+  int sampleSize_;
```
Comment (yu239) [inline paddle/gserver/layers/RotateLayer.h]: fixed

File: paddle/math/Matrix.cpp [hunk 6]
```diff
@@ -371,7 +383,7 @@ MatrixPtr GpuMatrix::getTranspose() {
...
 
-void GpuMatrix::transpose(MatrixPtr matTrans, bool memAlloc) {
+void GpuMatrix::transpose(MatrixPtr& matTrans, bool memAlloc) {
   if (memAlloc) {
```
Comment (yu239) [inline paddle/math/Matrix.cpp]: Yes. When memAlloc is true, the pointer matTrans will change, and thus the reference should be passed so that outside the function the pointer is also changed.

File: paddle/math/Matrix.cpp [hunk 7]
```diff
@@ -385,13 +397,29 @@ void GpuMatrix::transpose(MatrixPtr matTrans, bool memAlloc) {
   hl_matrix_transpose(data, dataTrans, height_, width_, lda, ldc);
 }
```
Comment (yu239) [inline paddle/math/Matrix.cpp]: See the header file Matrix.h

File: paddle/math/Matrix.cpp [hunk 8]
```diff
@@ -385,13 +397,29 @@ void GpuMatrix::transpose(MatrixPtr matTrans, bool memAlloc) {
...
+void GpuMatrix::rotate(MatrixPtr& matRot, bool memAlloc, bool clockWise) {
+  if (memAlloc) {
+    matRot = std::make_shared<GpuMatrix>(width_, height_);
+  } else {
+    CHECK(matRot != NULL);
+  }
+
+  MatrixPtr cpuMat = std::make_shared<CpuMatrix>(height_, width_);
+  cpuMat->copyFrom(*this);
+
```
Comment (yu239) [inline paddle/math/Matrix.cpp]: Added a GPU implementation

File: paddle/math/Matrix.cpp (hunk 3 above)
Comment (yu239) [inline paddle/math/Matrix.cpp]: Added

File: paddle/math/tests/test_SparseMatrix.cpp (hunk 1 above)
Comment (yu239) [inline paddle/math/tests/test_SparseMatrix.cpp]: Because I changed the first argument of transpose from value to reference, it cannot automatically cast the cpumatrix to the base matrix class anymore, so I have to cast it explicitly.

File: paddle/math/tests/test_matrixCompare.cpp (hunk 2 above)
Comment (yu239) [inline paddle/math/tests/test_matrixCompare.cpp]: Inverse also has the same problem with transpose when memAlloc is true. I want test here.

File: paddle/math/CpuSparseMatrix.cpp [hunk 9]
```diff
@@ -372,7 +372,7 @@ MatrixPtr CpuSparseMatrix::subMatrix(size_t startRow, size_t numRows) {
 }
 
 /* mem MUST be alloced outside (memAlloc=false) */
-void CpuSparseMatrix::transpose(MatrixPtr matTrans, bool memAlloc) {
+void CpuSparseMatrix::transpose(MatrixPtr& matTrans, bool memAlloc) {
```
Comment (wangkuiyi) [inline paddle/math/CpuSparseMatrix.cpp]: 我以为 MatrixPtr 已经是一个 pointer 了，下文里貌似也没有重新赋值给 matTrans，是不是不需要加 `&`?

File: paddle/math/Matrix.cpp [hunk 10]
```diff
@@ -1708,13 +1738,35 @@ void CpuMatrix::transpose(MatrixPtr matTrans, bool memAlloc) {
   }
 }
 
+void CpuMatrix::rotate(MatrixPtr& matRot, bool memAlloc, bool clockWise) {
```
Comment (wangkuiyi) [inline paddle/math/Matrix.cpp]: 这里我也以为只需要  MatrixPtr ，而不需要 \& 呢？

**Comment by wangkuiyi** [comment]
https://google.github.io/styleguide/cppguide.html#Reference_Arguments
